src/service/resources-impl.js

```diff
@@ -256,22 +256,49 @@ export class Resources {
   }
 
   /**
-   * Returns a subset of resources which is identified as being in the current
-   * viewport.
+   * Returns a subset of resources which are (1) belong to the specified host
+   * window, and (2) positioned in the specified rect.
+   * @param {!Window} hostWin
+   * @param {!../layout-rect.LayoutRectDef} rect
    * @param {boolean=} opt_isInPrerender signifies if we are in prerender mode.
-   * @return {!Array<!Resource>}
+   * @return {!Promise<!Array<!Resource>>}
    */
-  getResourcesInViewport(opt_isInPrerender) {
+  getResourcesInRect(hostWin, rect, opt_isInPrerender) {
     opt_isInPrerender = opt_isInPrerender || false;
-    const viewportRect = this.viewport_.getRect();
-    return this.resources_.filter(r => {
-      if (r.hasOwner() || !r.isDisplayed() || !r.overlaps(viewportRect)) {
-        return false;
-      }
-      if (opt_isInPrerender && !r.prerenderAllowed()) {
-        return false;
-      }
-      return true;
+
+    // First, wait for the `ready-scan` signal. Waiting for each element
+    // individually is too expensive and `ready-scan` will cover most of
+    // the initially parsed elements.
+    // TODO(jridgewell): this path should be switched to use a future
+    // "layer has been measured" signal.
+    return this.ampdoc.signals().whenSignal('ready-scan').then(() => {
```

Comment:
shall we put all signals into an enum and have a centralized place for documentation?